subsys/bluetooth/services/bas/bas_bls.c

```diff
@@ -0,0 +1,183 @@
+/*
+ * Copyright (c) 2024 Demant A/S
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "bas_internal.h"
+#include <zephyr/bluetooth/services/bas.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_DECLARE(bas, CONFIG_BT_BAS_LOG_LEVEL);
+
+static struct bt_bas_bls *battery_level_status;
+
+/*
+ * Bitfield structure: Power State
+ *
+ * - Bits 0: Battery Present
+ * - Bits 1–2: Wired External Power Source Connected
+ * - Bits 3–4: Wireless External Power Source Connected
+ * - Bits 5–6: Battery Charge State
+ * - Bits 7–8: Battery Charge Level
+ * - Bits 9–11: Charging Type
+ * - Bits 12–14: Charging Fault Reason
+ * - Bit 15: RFU
+ *
+ * For detailed specification, refer to:
+ * https://bitbucket.org/bluetooth-SIG/public/src/main/gss/
+ * org.bluetooth.characteristic.battery_level_status.yaml
+ */
+#define BATTERY_SHIFT              0
+#define WIRED_POWER_SHIFT          1
+#define WIRELESS_POWER_SHIFT       3
+#define BATTERY_CHARGE_STATE_SHIFT 5
+#define BATTERY_CHARGE_LEVEL_SHIFT 7
+#define BATTERY_CHARGE_TYPE_SHIFT  9
+#define CHARGING_FAULT_SHIFT       12
+
+/*
+ * Bitfield structure: Additional Status
+ *
+ * - Bits 0–1: Service Required
+ * - Bit 2: Battery Fault
+ * - Bits 3–7: Reserved
+ */
+#define SERVICE_REQUIRED_SHIFT 0
+#define BATTERY_FAULT_SHIFT    2
+
+void bt_bas_bls_init(struct bt_bas_bls *bls)
+{
+	LOG_INF("Initialise BAS Battery Level Status Module");
+
+	battery_level_status = bls;
+	if (battery_level_status == NULL) {
+		LOG_INF("BAS service not initialised");
+		return;
+	}
+
+	battery_level_status->flags = 0;
+	battery_level_status->power_state = 0;
+
+#if defined(CONFIG_BT_BAS_BLS_IDENTIFIER_PRESENT)
+	/* Set identifier flag */
+	battery_level_status->flags |= BIT(0);
+	battery_level_status->identifier = 0;
+#endif
+
+#if defined(CONFIG_BT_BAS_BLS_BATTERY_LEVEL_PRESENT)
+	/* Set battery level flag */
+	battery_level_status->flags |= BIT(1);
+	battery_level_status->battery_level = bt_bas_get_battery_level();
+#endif
+
+#if defined(CONFIG_BT_BAS_BLS_ADDITIONAL_STATUS_PRESENT)
+	/* Set additional status flag */
+	battery_level_status->flags |= BIT(2);
+	battery_level_status->additional_status = 0;
+#endif
+}
+
+void bt_bas_bls_set_battery_present(enum bt_bas_bls_battery_present present)
+{
+	if (battery_level_status == NULL) {
+		return;
+	}
```

Comment:
Hmm, if the functions can fail for a reason like this, should we have a return value for each of them?